src/Conversion/KrnlToAffine/KrnlToAffine.cpp

```diff
@@ -25,6 +25,215 @@
+// reference %loop_ref, we will need to maintain the relative positioning of IR
+// block and their parent loop operations; we do so by moving IR blocks while
+// Krnl Dialect lowering proceeds.
+//
+// Consider the following example, where we specify the recepie for a
+// 2-dimensional tiled loop, and insert memory allocation/deallocation aimed to
+// set up and clean up per-tile temporary buffer:
+//
+// %ii, %ij = krnl.define_loops 2
+// %ib, %il = krnl.block %ii 5 : (!krnl.loop) -> (!krnl.loop, !krnl.loop)
+// %jb, %jl = krnl.block %ij 4 : (!krnl.loop) -> (!krnl.loop, !krnl.loop)
+// krnl.permute(%ib, %il, %jb, %jl) [0, 2, 1, 3] : !krnl.loop, !krnl.loop,
+//     !krnl.loop, !krnl.loop
+// krnl.iterate(%ib, %jb) with (%ii -> %i = 0 to 10, %ij -> %j = 0 to 20) {
+//   %alloc = alloc() : memref<10 x f32>
+//   krnl.iterate(%il, %jl) with () {
+//     %foo = addi %i, %j : index
+//   }
+//   dealloc %alloc : memref<10 x f32>
+//  }
+//
+// The temporary buffer allocation/deallocation are placed within loops that
+// have yet to be materialized because loop tiling and loop permutation are only
+// specified as recipes without actually being applied at Krnl Dialect level.
+// Therefore as we proceed to lower Krnl Dialect, there will be no place for
+// these (blocks of) operations to exist until the corresponding concrete outer
+// loops emerge, as a result of optimizations being applied. Upon materializing
+// such a loop, we will move these (blocks of) operations to the corresponding
+// regions in the newly created loops.
+//
+// We use LoopBody mover to:
+// - register, for each Krnl loop reference, blocks of operations
+//   that should be contained directly beneath the corresponding concrete loops
+//   as the moving plan in the beginning of the Krnl Dialect lowering.
+// - subsequently, when the concrete loops corresponding to the Krnl loop
+//   reference is materialized, IR blocks will be moved to appropriate locations
+//   based on information recorded as moving plan.
+//
+// Thus, for the above IR, the following moving plan will be registered:
+// - For %ib, %jb, the list of operation nested directly under is:
+//    - alloc() operation,
+//    - materialized loops corresponding to %il, %jl,
+//    - dealloc() operation.
+// - For %il, %jl, the list of operations nested directly under is:
+//    - addi operation.
+//
+// Subsequently, lowering will start with affine ops materialized corresponding
+// to the reference to un-optimized loops:
+//
+// affine.for %i = 0 to 10 {
+//   affine.for %j = 0 to 20 {
+//     %foo = addi %i, %j : index
+//   }
+// }
+//
+// Since the tiling has not taken place yet, tile coordinat iteration loops have
+// not been materialized, therefore the alloc and dealloc operations do not fit
+// in the IR presently yet. Instead, they will be placed within a krnl.movable
+// op region, to indicate that their positioning is subject to change.
```

Comment:
Again, thanks for a very thorough description. If my understanding is correct, the movable is never generated by the user (aka the person writing the lowering of ONNX to KRNL), it is generated during the process of lowering. I would add this "This operation is automatically generated by the lowering of KRNL to affine to assist with patterns found, for example, during the tiling of imperfectly nested loops".

This mechanism is very ingenious.

Reply:
Yes. Will do.